File: src/bot/commands/player/units.ts
import { SlashCommandBuilder, type ChatInputCommandInteraction } from 'discord.js';
import { Command } from '../../struct/Command.js';
import { baseEmbed } from '../../../util/embeds.js';
import { EMOJIS } from '../../../util/emojis.js';
import { pad, padStart } from '../../../util/helper.js';

export default class UnitsCommand extends Command {
  public constructor() {
    super('units', { category: 'player', description: "Show a player's troop, spell and hero levels.", defer: true });
  }

  public builder() {
    return new SlashCommandBuilder()
      .setName('units')
      .setDescription("Show a player's troop, spell and hero levels.")
      .addStringOption((opt) => opt.setName('tag').setDescription('Player tag. Defaults to your linked account.'))
      .addUserOption((opt) => opt.setName('user').setDescription('Show another linked user.'));
  }

  private table(items: { name: string; level: number; maxLevel: number }[]): string {
    if (!items.length) return '_none_';
    return items
      .map((u) => `\`${pad(u.name, 18)} ${padStart(u.level, 2)}/${padStart(u.maxLevel, 2)}\``)
      .join('\n')
      .slice(0, 1024);
  }

  public async exec(interaction: ChatInputCommandInteraction, args: Record<string, any>) {
    const { data: player, error } = await this.client.resolver.resolvePlayer(interaction, args.tag, args.user);
    if (!player) return interaction.editReply(error!);

    const home = <T extends { village: string }>(arr: T[]) => arr.filter((x) => x.village === 'home');
    const pets = ((player as any).pets ?? []) as { name: string; level: number; maxLevel: number }[];

    const embed = baseEmbed(this.client, interaction).setTitle(`${player.name} (${player.tag}) — Units`);
    embed.addFields({ name: `${EMOJIS.HERO} Heroes`, value: this.table(home(player.heroes)) });
    if (pets.length) embed.addFields({ name: `${EMOJIS.PET} Pets`, value: this.table(pets) });
    embed.addFields(
      { name: `${EMOJIS.TROOP} Troops`, value: this.table(home(player.troops)) },
      { name: `${EMOJIS.SPELL} Spells`, value: this.table(home(player.spells)) }
    );

    return interaction.editReply({ embeds: [embed] });
  }
}

File: cogs/player/commands/units.py
"""/player units — troop, spell and hero levels."""

import discord

from utils.embeds import base_embed, error_embed
from utils.emojis import E_HERO, E_SPELL, E_TROOP
from utils.helpers import pad, pad_start
from utils.resolver import resolve_player


def _table(items) -> str:
    if not items:
        return "_none_"
    lines = [f"`{pad(u.name, 18)} {pad_start(u.level, 2)}/{pad_start(u.max_level, 2)}`" for u in items]
    return "\n".join(lines)[:1024]


async def units(interaction: discord.Interaction, tag: str = None, user: discord.User = None):
    await interaction.response.defer()
    try:
        player = await resolve_player(interaction, tag, user)
    except ValueError as err:
        await interaction.followup.send(embed=error_embed(str(err)))
        return

    embed = await base_embed(interaction, title=f"{player.name} ({player.tag}) — Units")
    embed.add_field(name=f"{E_HERO} Heroes", value=_table([h for h in player.heroes if h.is_home_base]), inline=False)
    if player.pets:
        embed.add_field(name="🐾 Pets", value=_table(player.pets), inline=False)
    embed.add_field(name=f"{E_TROOP} Troops", value=_table(player.home_troops), inline=False)
    embed.add_field(name=f"{E_SPELL} Spells", value=_table(player.spells), inline=False)

    await interaction.followup.send(embed=embed)
